core/divelist.c

```diff
@@ -259,6 +236,100 @@ static int calculate_cns(struct dive *dive)
 		j--;
 		cns += ((double)t) / ((double)cns_table[j][1]) * 100;
 	}
+	
+	return cns;
+}
+
+/* this only gets called if dive->maxcns == 0 which means we know that
+ * none of the divecomputers has tracked any CNS for us
+ * so we calculated it "by hand" */
+static int calculate_cns(struct dive *dive)
+{
+	int i, divenr;
+	double cns = 0.0;
+	timestamp_t last_starttime, last_endtime = 0;
+
+	/* shortcut */
+	if (dive->cns)
+		return dive->cns;
+
+	divenr = get_divenr(dive);
+	i = divenr >= 0 ? divenr : dive_table.nr;
+	if (i >= 0 && i < dive_table.nr) printf("\n\n*** CNS for dive #%d %d\n", i, get_dive(i)->number);
+	else printf("\n\n*** CNS for dive #%d\n", i);
+	/* Look at next dive in dive list table and correct i when needed */
+	while (i < dive_table.nr - 1) {
+		struct dive *pdive = get_dive(i);
+		if (!pdive || pdive->when > dive->when)
+			break;
+		i++;
+	}
+	/* Look at previous dive in dive list table and correct i when needed */
+	while (i > 0) {
+		struct dive *pdive = get_dive(i - 1);
+		if (!pdive || pdive->when < dive->when)
+			break;
+		i--;
+	}
```

Comment:
given that `sort_table()` uses `when` from the `struct dive`, do you think that the second iteration is still needed? technically if the list is sorted, you would know where `dive` fits chronologically.
is the list not sorted at this point in time?

Reply:
In the planner when you plan a new dive or replan an existing planned dive you will always have a dive which is "floating". What I mean is that the dive may have a dive table number but this number can be incorrect because the user can move the date of the dive around. At this moment the list is not re-sorted all the time. Re-sorting is done not before exiting the planner.
At the same moment we need to identify potential previous dives for this "floating" dive. This is why we need to identify something like a temporary dive table number.